proposals/list-patterns.md

```diff
@@ -0,0 +1,122 @@
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is {1, 2, 3}` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+The pattern syntax is modified as follow:
+
+```antlr
+recursive_pattern
+  : type? positional_pattern_clause? length_pattern_clause? property_or_list_pattern_clause? simple_designation?
+  ;
+
+length_pattern_clause
+  : '[' pattern ']'
+  ;
+
+property_or_list_pattern_clause
+  : list_pattern_clause
+  | property_pattern_clause
+  ;
+
+property_pattern_clause
+  : '{' (subpattern (',' subpattern)* ','?)? '}'
+  ;
+
+list_pattern_clause
+  : '{' pattern (',' pattern)* ','? '}'
+  ;
+
+slice_pattern
+  : '..' negated_pattern?
+  ;
+
+primary_pattern
+  : recursive_pattern
+  | slice_pattern
+  ;
+```
+There are two new additions to the *recursive_pattern* syntax as well as a *slice_pattern*:
+
+- The *list_pattern_clause* is used to match elements and the *length_pattern_clause* is used to match the length.
+- A *slice_pattern* is only permitted once and only directly in a *list_pattern_clause* and discards _**zero or more**_ elements.
+
+Due to the ambiguity with *property_pattern_clause*, the *list_pattern_clause* cannot be empty and a *length_pattern_clause* should be used instead.
+
+#### Pattern compatibility
+
+A *length_pattern_clause* is compatible with any type that conforms to the following rules:
+
+1. Has an accessible property getter that returns an `int` and has the name `Length` or `Count`
+
+A *list_pattern_clause* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the length pattern
+2. Has an accessible indexer with a single `int` parameter
+
+A *slice_pattern* is compatible with any type that conforms to the following rules:
```

Comment:
What is the type of a slice pattern? (what's the `var` in `..var x`?)

Reply:
That's just a var pattern. Currently that's the only sensible pattern here, but I think user-defined positional patterns (https://github.com/dotnet/csharplang/issues/4131) should be useful here in the future.

Reply:
Sorry I didn't follow. How is `..var x` just a var pattern? It has `..` for slicing.

Reply:
Given:
```antlr
slice_pattern
  : '..' negated_pattern?
  ;
```
`.. var x` is `..` followed by a var pattern which is matched against the slice (just capturing it in this case)

You could imagine a user-defined pattern against `Span` so that you could write:

```cs
expr is {1, .. SomePattern(), 3}
```
```cs
expr.Length    is >= 2
&& expr[0]     is 1
&& expr[1..^1] is SomePattern()
&& expr[^1]    is 3
```

Reply:
Let me illustrate the question with an example:
`x is { 1, .. var { Property: 0 } y, 2 }` // if this is allowed, what is the type of `var` and `y`, and what is the receiver type for `Property`?

Reply:
`var { Property: 0 } y` is not a valid pattern anywhere. 

Note that `var` is not part of the slice pattern, rather it's a "pattern combinator" so that it could contain other patterns in it.

Reply:
Sorry my example was invalid. We can break it into two separate patterns `..var y` and  `..{ Property: 0 } y`, and my question stands.

Reply:
>We'll want to say what is the input type for patterns following `..`. It's whatever type the range indexer returned.

Correct. I'll add a note to clarify that.